src/MergeContext.js

```diff
@@ -337,19 +355,55 @@ class MergeContext {
         const userRequested = usersVoted.find(el => el.state === 'changes_requested');
         if (userRequested !== undefined) {
             this._log("changes requested by " + userRequested.reviewer);
-            return null;
+            this._approvalResult.assign("blocked (see change requests)", "pending", null);
+            return;
         }
         const usersApproved = usersVoted.filter(u => u.state !== 'changes_requested');
         this._log("approved by " + usersApproved.length + " core developer(s)");
 
         if (usersApproved.length < Config.necessaryApprovals()) {
             this._log("not approved by necessary " + Config.necessaryApprovals() + " votes");
-            return null;
+            this._approvalResult.assign("waiting for more votes", "pending", null);
+            return;
+        }
+
+        const prAgeMs = new Date() - new Date(this._createdAt());
+        if (prAgeMs < Config.votingDelayMin()) {
+            this._approvalResult.assign("waiting for fast track objections", "pending", Config.votingDelayMin() - prAgeMs);
+            return;
+        }
+
+        if (usersApproved.length >= Config.sufficientApprovals()) {
+            this._approvalResult.assign("approved", "success", 0);
+            return;
+        }
+
+        if (prAgeMs >= Config.votingDelayMax()) {
+            this._approvalResult.assign("approved (on slow burner)", "success", 0);
+            return;
+        }
+        this._approvalResult.assign("waiting for more votes or a slow burner timeout", "pending", Config.votingDelayMax() - prAgeMs);
+        return;
+    }
+
+    async _setApprovalStatus(sha) {
+        if (this._dryRun("setting approval statuses"))
+            return;
+        if (!Config.manageApprovalStatus())
+            return;
+
+        assert(this._approvalResult);
+        let combinedStatus = await GH.getStatuses(sha);
+        const approvalStatus = combinedStatus.statuses ?
+            combinedStatus.statuses.find(el => el.context.trim() === Config.approvalContext()) : null;
+
+        if (approvalStatus && approvalStatus.state === this._approvalResult.state &&
```

Comment:
I would move the two comparisons into Approval::matchesGitHubStatusCheck() or similar. As a bonus, that method can check that `this.state` and `this.description` are not null.